sklearn/metrics/_regression.py

```diff
@@ -608,13 +609,75 @@ def median_absolute_error(
     return np.average(output_errors, weights=multioutput)
 
 
+def _assemble_r2_explained_variance(
+    numerator, denominator, n_outputs, multioutput, force_finite
+):
+    """Common part used by explained variance score and :math:`R^2` score."""
+
+    nonzero_denominator = denominator != 0
+
+    if not force_finite:
+        # Standard formula, that may lead to NaN or -Inf
+        output_scores = 1 - (numerator / denominator)
+    else:
+        nonzero_numerator = numerator != 0
+        # Default = Zero Numerator = perfect predictions. Set to 1.0
+        # (note: even if denominator is zero, thus avoiding NaN scores)
+        output_scores = np.ones([n_outputs])
+        # Non-zero Numerator and Non-zero Denominator: use the formula
+        valid_score = nonzero_denominator & nonzero_numerator
+        output_scores[valid_score] = 1 - (
+            numerator[valid_score] / denominator[valid_score]
+        )
+        # Non-zero Numerator and Zero Denominator:
+        # arbitrary set to 0.0 to avoid -inf scores
+        output_scores[nonzero_numerator & ~nonzero_denominator] = 0.0
+
+    if isinstance(multioutput, str):
+        if multioutput == "raw_values":
+            # return scores individually
+            return output_scores
+        elif multioutput == "uniform_average":
+            # Passing None as weights to np.average results is uniform mean
+            avg_weights = None
+        elif multioutput == "variance_weighted":
+            avg_weights = denominator
+            if not np.any(nonzero_denominator):
+                # All weights are zero, np.average would raise a ZeroDiv error.
+                # This only happens when all y are constant (or 1-element long)
+                # Since weights are all equal, fall back to uniform weights.
+                avg_weights = None
```

Comment:
@glemaitre  here is the new version of the code you commented for non-coverage. The results of this version are identical with the ones of the previous version, I checked.

Reply:
OK so let see if the coverage will be happy ;)